src/main/java/io/confluent/connect/hdfs/wal/QFSWAL.java

```diff
@@ -0,0 +1,230 @@
+  private final Pattern pattern;
+  private final String lockDir;
+  private String lockFile;
+
+  private final int lockRefreshIntervalInMs;
+  private final int lockTimeoutInMs;
+  private static final Logger log = LoggerFactory.getLogger(QFSWAL.class);
+
+  public QFSWAL(String logsDir, TopicPartition topicPartition, HdfsStorage storage) {
+    this(logsDir,
+         topicPartition,
+         storage,
+         DEFAULT_LOCK_REFRESH_INTERVAL_IN_MS,
+         DEFAULT_LOCK_TIMEOUT_IN_MS);
+  }
+
+  public QFSWAL(
+          String logsDir,
+          TopicPartition topicPartition,
+          HdfsStorage storage,
+          int lockRefreshIntervalInMs,
+          int lockTimeoutInMs
+  ) {
+    this.uuid = UUID.randomUUID();
+    this.pattern = Pattern.compile(String.format(
+            "([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})-(\\d+)(\\.)%s",
+            LOCK_FILE_EXTENSION));
+    this.timer = new Timer("QFSWAL-Timer", true);
+    this.storage = storage;
+    this.logsDir = logsDir;
+    this.topicPartition = topicPartition;
+    this.lockDir = FileUtils.directoryName(storage.url(), logsDir, topicPartition);
+    this.lockRefreshIntervalInMs = lockRefreshIntervalInMs;
+    this.lockTimeoutInMs = lockTimeoutInMs;
+    this.lockFile = this.createOrRenameLockFile();
+    this.startRenamingTimer();
+  }
+
+  private String getNewLockFile() {
+    return FileUtils.fileName(storage.url(),
+                              this.logsDir,
+                              this.topicPartition,
+                              String.format("%s-%s.%s",
+                                            this.uuid,
+                                            System.currentTimeMillis(),
+                                            LOCK_FILE_EXTENSION));
+  }
+
+  private String createOrRenameLockFile() {
+    if (!this.storage.exists(this.lockDir)) {
+      this.storage.create(this.lockDir);
+    }
+
+    List<SimpleEntry<String, Matcher>> lockFiles = this.findLockFiles();
+    String newLockFile = getNewLockFile();
+
+    long liveLocksCount = lockFiles.stream()
+            .filter(pair -> Long.parseLong(
+                    pair.getValue().group(2))
+                    > System.currentTimeMillis() - this.lockTimeoutInMs
+             ).count();
+
+    if (liveLocksCount != 0) {
+      throw new ConnectException("Lock has been acquired by another process");
+    }
+
+    if (lockFiles.isEmpty()) {
+      this.storage.create(newLockFile, true);
+    } else {
+      String oldLockFile = FileUtils.fileName(storage.url(),
+                                              this.logsDir,
+                                              this.topicPartition,
+                                              lockFiles.get(0).getKey());
+      this.storage.commit(oldLockFile, newLockFile);
+    }
+
+    liveLocksCount = this.findLockFiles()
+            .stream()
+            .filter(pair -> Long.parseLong(
+                    pair.getValue().group(2))
```

Comment:
Can you use named groups for pulling out this

Reply:
Actually I think findLockFiles should really be returning a proper object called LockFile which has the values already parsed out in one place. It could use named regex groups to make that bit easier to digest.


something like 
```
static class LockFile {
  LockFile(Matcher, String) {
  ...
  }

  final UUID uuid;
  final Instant instant;
  final String path;
}
```